### ISSUE TITLE
[Data] Support automatic batching in iter_torch_batches for all data types

### ISSUE THREAD

**Issue by woshiyyya** [ray-project/ray#40960]
### Description

The default batch collation function in iter_torch_batches automatically detects the data type and creates pytorch tensors for a batch of instances. This is quite convenient, but only works if all columns are numeric data types:

https://github.com/ray-project/ray/blob/c3285a8f9ae4d04c25bf16f6d87e4a9ac13e420b/python/ray/air/_internal/torch_utils.py#L146-L152

However, a more realistic case is: A user has multiple columns of data in a batch, but not all of them are numerical types. For example, the input sentences and labels could be str type.

Currently, users have to implement their own collate function if there are other data types (e.g. str) in a sample. This requires a lot of extra work. To simplify this process, we can change the behavior of the default collate function as follows:
- For numerical data types, we will automatically batch them.
- **For non-numerical types, we will simply return a list of objects.**

And also, this is exactly what PyTorch DataLoader does:

```python
from torch.utils.data import Dataset, DataLoader

# Example list of dictionaries
data_list = [
    {'a': np.array([0, 1]), 'b': 0, 'c': "adws"},
    {'a': np.array([1, 2]), 'b': 1, 'c': "kadw"},
]

# Define the dataset class
class DictDataset(Dataset):
    def __init__(self, data_list):
        self.data_list = data_list
    
    def __len__(self):
        return len(self.data_list)
    
    def __getitem__(self, idx):
        return self.data_list[idx]

# Example usage with DataLoader
dataset = DictDataset(data_list)
dataloader = DataLoader(dataset, batch_size=2, shuffle=True)

# Iterate through the dataset
for batch in dataloader:
    print(f"batch: {batch}")
```

The output batch would be:
```
Features: {'a': tensor([[1, 2], [0, 1]]), 'b': tensor([1, 0]), 'c': ['kadw', 'adws']}
```

This change would significantly increase the flexibility of Ray Data Ingestion. Users can provide arbitrary data types in a batch, without having to re-implement a collate batching function from time to time.
